File: twidere/src/main/java/org/mariotaku/twidere/util/ErrorInfoStore.java
package org.mariotaku.twidere.util;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.mariotaku.twidere.R;
import org.mariotaku.twidere.model.UserKey;


/**
 * Created by mariotaku on 16/1/31.
 */
public class ErrorInfoStore {

    public static final String KEY_DIRECT_MESSAGES = "direct_messages";
    public static final String KEY_INTERACTIONS = "interactions";
    public static final String KEY_HOME_TIMELINE = "home_timeline";
    public static final String KEY_ACTIVITIES_BY_FRIENDS = "activities_by_friends";

    public static final int CODE_NO_DM_PERMISSION = 1;
    public static final int CODE_NO_ACCESS_FOR_CREDENTIALS = 2;
    public static final int CODE_NETWORK_ERROR = 3;

    private final SharedPreferences mPreferences;

    public ErrorInfoStore(Application application) {
        mPreferences = application.getSharedPreferences("error_info", Context.MODE_PRIVATE);
    }

    public int get(String key) {
        return mPreferences.getInt(key, 0);
    }

    public int get(String key, String extraId) {
        return get(key + "_" + extraId);
    }

    public int get(String key, UserKey extraId) {
        final String host = extraId.getHost();
        if (host == null) {
            return get(key, extraId.getId());
        } else {
            return get(key + "_" + extraId.getId() + "_" + host);
        }
    }

    public void put(String key, int code) {
        mPreferences.edit().putInt(key, code).apply();
    }

    public void put(String key, String extraId, int code) {
        put(key + "_" + extraId, code);
    }

    public void put(String key, UserKey extraId, int code) {
        final String host = extraId.getHost();
        if (host == null) {
            put(key, extraId.getId(), code);
        } else {
            put(key + "_" + extraId.getId() + "_" + host, code);
        }
    }

    @Nullable
    public static DisplayErrorInfo getErrorInfo(@NonNull Context context, int code) {
        switch (code) {
            case CODE_NO_DM_PERMISSION: {
                return new DisplayErrorInfo(code, R.drawable.ic_info_error_generic,
                        context.getString(R.string.error_no_dm_permission));
            }
            case CODE_NO_ACCESS_FOR_CREDENTIALS: {
                return new DisplayErrorInfo(code, R.drawable.ic_info_error_generic,
                        context.getString(R.string.error_no_access_for_credentials));
            }
            case CODE_NETWORK_ERROR: {
                return new DisplayErrorInfo(code, R.drawable.ic_info_error_generic,
                        context.getString(R.string.network_error));
            }
        }
        return null;
    }

    public void remove(String key, String extraId) {
        remove(key + "_" + extraId);
    }

    public void remove(String key, UserKey extraId) {
        final String host = extraId.getHost();
        if (host == null) {
            remove(key, extraId.getId());
        } else {
            remove(key + "_" + extraId.getId() + "_" + host);
        }
    }

    public void remove(String key) {
        mPreferences.edit().remove(key).apply();
    }

    public static class DisplayErrorInfo {
        int code;
        int icon;
        String message;

        public DisplayErrorInfo(int code, int icon, String message) {
            this.code = code;
            this.icon = icon;
            this.message = message;
        }

        public int getCode() {
            return code;
        }

        public int getIcon() {
            return icon;
        }

        public String getMessage() {
            return message;
        }
    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/util/ErrorInfoStore.kt
package org.mariotaku.twidere.util

import android.content.Context
import androidx.annotation.DrawableRes
import org.mariotaku.twidere.R
import org.mariotaku.twidere.model.UserKey


/**
 * Created by mariotaku on 16/1/31.
 */
class ErrorInfoStore(application: Context) {

    private val preferences = application.getSharedPreferences("error_info", Context.MODE_PRIVATE)

    operator fun get(key: String): Int {
        return preferences.getInt(key, 0)
    }

    operator fun get(key: String, extraId: String): Int {
        return get(key + "_" + extraId)
    }

    operator fun get(key: String, extraId: UserKey): Int {
        val host = extraId.host
        return if (host == null) {
            get(key, extraId.id)
        } else {
            get(key + "_" + extraId.id + "_" + host)
        }
    }

    operator fun set(key: String, code: Int) {
        preferences.edit().putInt(key, code).apply()
    }

    operator fun set(key: String, extraId: String, code: Int) {
        set(key + "_" + extraId, code)
    }

    operator fun set(key: String, extraId: UserKey, code: Int) {
        val host = extraId.host
        if (host == null) {
            set(key, extraId.id, code)
        } else {
            set(key + "_" + extraId.id + "_" + host, code)
        }
    }

    fun remove(key: String, extraId: String) {
        remove(key + "_" + extraId)
    }

    fun remove(key: String, extraId: UserKey) {
        val host = extraId.host
        if (host == null) {
            remove(key, extraId.id)
        } else {
            remove(key + "_" + extraId.id + "_" + host)
        }
    }

    fun remove(key: String) {
        preferences.edit().remove(key).apply()
    }

    class DisplayErrorInfo(code: Int, @DrawableRes icon: Int, message: String) {
        var code: Int = 0
            internal set
        var icon: Int = 0
            internal set
        var message: String
            internal set

        init {
            this.code = code
            this.icon = icon
            this.message = message
        }
    }

    companion object {

        const val KEY_DIRECT_MESSAGES = "direct_messages"
        const val KEY_INTERACTIONS = "interactions"
        const val KEY_HOME_TIMELINE = "home_timeline"
        const val KEY_ACTIVITIES_BY_FRIENDS = "activities_by_friends"

        const val CODE_NO_DM_PERMISSION = 1
        const val CODE_NO_ACCESS_FOR_CREDENTIALS = 2
        const val CODE_NETWORK_ERROR = 3
        const val CODE_TIMESTAMP_ERROR = 4

        fun getErrorInfo(context: Context, code: Int): DisplayErrorInfo? {
            when (code) {
                CODE_NO_DM_PERMISSION -> {
                    return DisplayErrorInfo(code, R.drawable.ic_info_error_generic,
                            context.getString(R.string.error_no_dm_permission))
                }
                CODE_NO_ACCESS_FOR_CREDENTIALS -> {
                    return DisplayErrorInfo(code, R.drawable.ic_info_error_generic,
                            context.getString(R.string.error_no_access_for_credentials))
                }
                CODE_NETWORK_ERROR -> {
                    return DisplayErrorInfo(code, R.drawable.ic_info_error_generic,
                            context.getString(R.string.message_toast_network_error))
                }
                CODE_TIMESTAMP_ERROR -> {
                    return DisplayErrorInfo(code, R.drawable.ic_info_error_generic,
                            context.getString(R.string.error_info_oauth_timestamp_error))
                }
            }
            return null
        }
    }
}
